### PR TITLE
Add `RecentEraLedgerConstraints` for removing boilerplate

### PR DESCRIPTION
**Pull request by Anviking**
### Issue Number

Follow-up from #3881.

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 1]
```diff
@@ -958,40 +966,13 @@ evaluateTransactionBalance era pp utxo txBody' =
...
-
-withAlonzoScriptConstraint
+-- | Bring constraints related to 'IsRecentEra' into scope from a value-level
+-- 'RecentEra'.
+withConstraints
     :: RecentEra era
-    -> ( Core.Script (Cardano.ShelleyLedgerEra era)
-         ~ AlonzoScript (Cardano.ShelleyLedgerEra era)
-        => a)
+    -> (IsRecentEra era => a)
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]: Couldn't get rid of these helpers completely as we still take `RecentEra` on the value-level. Need for value-level `RecentEra` should be reducible with ADP-2353 though.

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 2]
```diff
@@ -398,21 +414,18 @@ modifyTxOutValue
...
         Alonzo.AlonzoTxOut addr (f val) dat
 
 modifyTxOutCoin
     :: RecentEra era
     -> (Coin -> Coin)
     -> TxOut (ShelleyLedgerEra era)
     -> TxOut (ShelleyLedgerEra era)
-modifyTxOutCoin era f = withStandardCryptoConstraint era $
+modifyTxOutCoin era f =
     modifyTxOutValue era (modifyCoin f)
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]:
```suggestion
modifyTxOutCoin era = modifyTxOutValue era . modifyCoin
```

File: lib/wallet/src/Cardano/Wallet/Write/Tx.hs [hunk 3]
```diff
@@ -439,26 +452,19 @@ scriptFromCardanoScriptInAnyLang
...
--- not be a problem.
 scriptToCardanoScriptInAnyLang
     :: forall era. IsRecentEra era
     => Script (Cardano.ShelleyLedgerEra era)
     -> Cardano.ScriptInAnyLang
-scriptToCardanoScriptInAnyLang = withAlonzoScriptConstraint (recentEra @era)
-    $ rewrap
+scriptToCardanoScriptInAnyLang =
+    rewrap
     . Cardano.fromShelleyBasedScript shelleyEra
```
Comment (jonathanknowles) [inline lib/wallet/src/Cardano/Wallet/Write/Tx.hs]:
```suggestion
scriptToCardanoScriptInAnyLang =
    rewrap . Cardano.fromShelleyBasedScript shelleyEra
```

**Review by jonathanknowles — APPROVED** [review]
Hi @Anviking 

I think overall this is an improvement.

It seems to me that we have a trade off between:
- minimal boilerplate, but some redundant constraints
- no redundant constraints, but large amounts of boilerplate

I think that in this case, I'm more swayed by the minimal boilerplate argument, because it makes the essence of the code much easier to read.

I only have a couple of small suggestions.

**Comment by Anviking** [comment]
Wow, there's some real CI errors I didn't see locally on 9.2:
```
src/Cardano/Wallet.hs:2913:15: error:
    <E2><80><A2> Couldn't match type <E2><80><98>m0<E2><80><99> with <E2><80><98>IO<E2><80><99>
        <E2><80><98>m0<E2><80><99> is untouchable
          inside the constraints: WriteTx.IsRecentEra era
          bound by a type expected by the context:
                     forall era.
                     WriteTx.IsRecentEra era =>
                     WriteTx.RecentEra era -> IO DelegationFee
          at src/Cardano/Wallet.hs:(2901,33)-(2916,54)
      Expected type: ExceptT ErrWalletNotInitialized m0 Coin
        Actual type: ExceptT ErrWalletNotInitialized IO Coin
    <E2><80><A2> In the second argument of <E2><80><98>($)<E2><80><99>, namely
        <E2><80><98>mapExceptT atomically isStakeKeyRegistered
           <&>
             \case
               False -> stakeKeyDeposit (Write.pparamsWallet protocolParams)
               True -> Coin 0<E2><80><99>
      In the second argument of <E2><80><98>($)<E2><80><99>, namely
        <E2><80><98>throwInIO . mkNoSuchWalletError walletId
           $ mapExceptT atomically isStakeKeyRegistered
               <&>
                 \case
                   False -> stakeKeyDeposit (Write.pparamsWallet protocolParams)
                   True -> Coin 0<E2><80><99>
      In a stmt of a 'do' block:
        deposit <- liftIO
                     $ throwInIO . mkNoSuchWalletError walletId
                         $ mapExceptT atomically isStakeKeyRegistered
                             <&>
                               \case
                                 False -> stakeKeyDeposit (Write.pparamsWallet protocolParams)
                                 True -> Coin 0
    <E2><80><A2> Relevant bindings include
        throwInIO :: forall c. ExceptT ErrNoSuchWallet m0 c -> m0 c
          (bound at src/Cardano/Wallet.hs:2918:5)
     |
2913 |             $ mapExceptT atomically isStakeKeyRegistered <&> \case
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...

src/Cardano/Wallet.hs:2918:17: error:
    <E2><80><A2> Could not deduce (Monad m0) arising from a use of <E2><80><98>>=><E2><80><99>
      from the context: (AddressBookIso s, WalletFlavor s n k)
        bound by the type signature for:
                   delegationFee :: forall s (k :: Depth -> * -> *)
                                           (n :: Cardano.Wallet.Read.NetworkId.NetworkDiscriminant).
                                    (AddressBookIso s, WalletFlavor s n k) =>
                                    DBLayer IO s k
                                    -> NetworkLayer IO Read.Block
                                    -> TransactionLayer k 'CredFromKeyK SealedTx
                                    -> TimeInterpreter (ExceptT PastHorizonException IO)
                                    -> AnyRecentEra
                                    -> ChangeAddressGen s
                                    -> WalletId
                                    -> IO DelegationFee
        at src/Cardano/Wallet.hs:(2887,1)-(2899,23)
      or from: (MonadIO stm, MonadFail stm)
        bound by a pattern with constructor:
                   DBLayer :: forall (m :: * -> *) s (k :: Depth -> * -> *)
                                     (stm :: * -> *).
                              (MonadIO stm, MonadFail stm) =>
                              WalletId
                              -> (Wallet s
                                  -> WalletMetadata
                                  -> [(Tx, TxMeta)]
                                  -> GenesisParameters
                                  -> ExceptT
                                       Cardano.Wallet.DB.Errors.ErrWalletAlreadyInitialized stm ())
                              -> Data.DBVar.DBVar
                                   stm
                                   (WalletState.DeltaMap WalletId (WalletState.DeltaWalletState s))
                              -> (Wallet s -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (Wallet s))
                              -> stm [ChainPoint]
                              -> (WalletMetadata -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (WalletMetadata, WalletDelegation))
                              -> ExceptT ErrWalletNotInitialized stm Bool
                              -> (DelegationCertificate
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Coin -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm Coin
                              -> ([(Tx, TxMeta)] -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Maybe Coin
                                  -> SortOrder
                                  -> Range SlotNo
                                  -> Maybe TxStatus
                                  -> Maybe Natural
                                  -> stm [TransactionInfo])
                              -> (Hash "Tx"
                                  -> ExceptT ErrWalletNotInitialized stm (Maybe TransactionInfo))
                              -> (BuiltTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx"
                                  -> SealedTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm
                                   [Cardano.Wallet.DB.Store.Submissions.Operations.TxSubmissionsStatus]
                              -> (SlotNo
                                  -> [(SlotNo, Hash "Tx")]
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx" -> ExceptT ErrRemoveTx stm ())
                              -> ((k 'RootK XPrv, PassphraseHash)
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (k 'RootK XPrv, PassphraseHash))
                              -> stm (Maybe GenesisParameters)
                              -> (Slot -> ExceptT ErrWalletNotInitialized stm ChainPoint)
                              -> (Quantity "block" GHC.Word.Word32
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (forall a. stm a -> m a)
                              -> DBLayer m s k,
                 in an equation for <E2><80><98>delegationFee<E2><80><99>
        at src/Cardano/Wallet.hs:2900:18-28
      The type variable <E2><80><98>m0<E2><80><99> is ambiguous
      Relevant bindings include
        throwInIO :: ExceptT ErrNoSuchWallet m0 c -> m0 c
          (bound at src/Cardano/Wallet.hs:2918:5)
      These potential instances exist:
        instance Monad (Either e) -- Defined in <E2><80><98>Data.Either<E2><80><99>
        instance Monad NonEmpty -- Defined in <E2><80><98>GHC.Base<E2><80><99>
        instance Monad IO -- Defined in <E2><80><98>GHC.Base<E2><80><99>
        ...plus 9 others
        ...plus 153 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    <E2><80><A2> In the expression:
        runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
      In an equation for <E2><80><98>throwInIO<E2><80><99>:
          throwInIO
            = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
      In an equation for <E2><80><98>delegationFee<E2><80><99>:
          delegationFee
            db@DBLayer {..}
            netLayer
            txLayer
            ti
            era
            changeAddressGen
            walletId
            = WriteTx.withRecentEra era
                $ \ (recentEra :: WriteTx.RecentEra era)
                    -> do protocolParams <- Write.unsafeFromWalletProtocolParameters
                                              <$> liftIO (currentProtocolParameters netLayer)
                          ....
            where
                throwInIO
                  = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |
2918 |     throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Cardano/Wallet.hs:2918:40: error:
    <E2><80><A2> Could not deduce (MonadIO m0) arising from a use of <E2><80><98>throwIO<E2><80><99>
      from the context: (AddressBookIso s, WalletFlavor s n k)
        bound by the type signature for:
                   delegationFee :: forall s (k :: Depth -> * -> *)
                                           (n :: Cardano.Wallet.Read.NetworkId.NetworkDiscriminant).
                                    (AddressBookIso s, WalletFlavor s n k) =>
                                    DBLayer IO s k
                                    -> NetworkLayer IO Read.Block
                                    -> TransactionLayer k 'CredFromKeyK SealedTx
                                    -> TimeInterpreter (ExceptT PastHorizonException IO)
                                    -> AnyRecentEra
                                    -> ChangeAddressGen s
                                    -> WalletId
                                    -> IO DelegationFee
        at src/Cardano/Wallet.hs:(2887,1)-(2899,23)
      or from: (MonadIO stm, MonadFail stm)
        bound by a pattern with constructor:
                   DBLayer :: forall (m :: * -> *) s (k :: Depth -> * -> *)
                                     (stm :: * -> *).
                              (MonadIO stm, MonadFail stm) =>
                              WalletId
                              -> (Wallet s
                                  -> WalletMetadata
                                  -> [(Tx, TxMeta)]
                                  -> GenesisParameters
                                  -> ExceptT
                                       Cardano.Wallet.DB.Errors.ErrWalletAlreadyInitialized stm ())
                              -> Data.DBVar.DBVar
                                   stm
                                   (WalletState.DeltaMap WalletId (WalletState.DeltaWalletState s))
                              -> (Wallet s -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (Wallet s))
                              -> stm [ChainPoint]
                              -> (WalletMetadata -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (WalletMetadata, WalletDelegation))
                              -> ExceptT ErrWalletNotInitialized stm Bool
                              -> (DelegationCertificate
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Coin -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm Coin
                              -> ([(Tx, TxMeta)] -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Maybe Coin
                                  -> SortOrder
                                  -> Range SlotNo
                                  -> Maybe TxStatus
                                  -> Maybe Natural
                                  -> stm [TransactionInfo])
                              -> (Hash "Tx"
                                  -> ExceptT ErrWalletNotInitialized stm (Maybe TransactionInfo))
                              -> (BuiltTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx"
                                  -> SealedTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm
                                   [Cardano.Wallet.DB.Store.Submissions.Operations.TxSubmissionsStatus]
                              -> (SlotNo
                                  -> [(SlotNo, Hash "Tx")]
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx" -> ExceptT ErrRemoveTx stm ())
                              -> ((k 'RootK XPrv, PassphraseHash)
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (k 'RootK XPrv, PassphraseHash))
                              -> stm (Maybe GenesisParameters)
                              -> (Slot -> ExceptT ErrWalletNotInitialized stm ChainPoint)
                              -> (Quantity "block" GHC.Word.Word32
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (forall a. stm a -> m a)
                              -> DBLayer m s k,
                 in an equation for <E2><80><98>delegationFee<E2><80><99>
        at src/Cardano/Wallet.hs:2900:18-28
      The type variable <E2><80><98>m0<E2><80><99> is ambiguous
      Relevant bindings include
        throwInIO :: ExceptT ErrNoSuchWallet m0 c -> m0 c
          (bound at src/Cardano/Wallet.hs:2918:5)
      These potential instances exist:
        instance [safe] MonadIO IO -- Defined in <E2><80><98>Control.Monad.IO.Class<E2><80><99>
        instance [safe] MonadIO m => MonadIO (ExceptT e m)
          -- Defined in <E2><80><98>Control.Monad.Trans.Except<E2><80><99>
        instance [safe] MonadIO m => MonadIO (MaybeT m)
          -- Defined in <E2><80><98>Control.Monad.Trans.Maybe<E2><80><99>
        ...plus one other
        ...plus 39 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    <E2><80><A2> In the first argument of <E2><80><98>(.)<E2><80><99>, namely <E2><80><98>throwIO<E2><80><99>
      In the first argument of <E2><80><98>either<E2><80><99>, namely
        <E2><80><98>(throwIO . ExceptionNoSuchWallet)<E2><80><99>
      In the second argument of <E2><80><98>(>=>)<E2><80><99>, namely
        <E2><80><98>either (throwIO . ExceptionNoSuchWallet) pure<E2><80><99>
     |
2918 |     throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |                                        ^^^^^^^

src/Cardano/Wallet.hs:2918:73: error:
    <E2><80><A2> Could not deduce (Applicative m0) arising from a use of <E2><80><98>pure<E2><80><99>
      from the context: (AddressBookIso s, WalletFlavor s n k)
        bound by the type signature for:
                   delegationFee :: forall s (k :: Depth -> * -> *)
                                           (n :: Cardano.Wallet.Read.NetworkId.NetworkDiscriminant).
                                    (AddressBookIso s, WalletFlavor s n k) =>
                                    DBLayer IO s k
                                    -> NetworkLayer IO Read.Block
                                    -> TransactionLayer k 'CredFromKeyK SealedTx
                                    -> TimeInterpreter (ExceptT PastHorizonException IO)
                                    -> AnyRecentEra
                                    -> ChangeAddressGen s
                                    -> WalletId
                                    -> IO DelegationFee
        at src/Cardano/Wallet.hs:(2887,1)-(2899,23)
      or from: (MonadIO stm, MonadFail stm)
        bound by a pattern with constructor:
                   DBLayer :: forall (m :: * -> *) s (k :: Depth -> * -> *)
                                     (stm :: * -> *).
                              (MonadIO stm, MonadFail stm) =>
                              WalletId
                              -> (Wallet s
                                  -> WalletMetadata
                                  -> [(Tx, TxMeta)]
                                  -> GenesisParameters
                                  -> ExceptT
                                       Cardano.Wallet.DB.Errors.ErrWalletAlreadyInitialized stm ())
                              -> Data.DBVar.DBVar
                                   stm
                                   (WalletState.DeltaMap WalletId (WalletState.DeltaWalletState s))
                              -> (Wallet s -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (Wallet s))
                              -> stm [ChainPoint]
                              -> (WalletMetadata -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (WalletMetadata, WalletDelegation))
                              -> ExceptT ErrWalletNotInitialized stm Bool
                              -> (DelegationCertificate
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Coin -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm Coin
                              -> ([(Tx, TxMeta)] -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Maybe Coin
                                  -> SortOrder
                                  -> Range SlotNo
                                  -> Maybe TxStatus
                                  -> Maybe Natural
                                  -> stm [TransactionInfo])
                              -> (Hash "Tx"
                                  -> ExceptT ErrWalletNotInitialized stm (Maybe TransactionInfo))
                              -> (BuiltTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx"
                                  -> SealedTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm
          -- Defined in <E2><80><98>Control.Monad.Trans.Maybe<E2><80><99>
        ...plus one other
        ...plus 39 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    <E2><80><A2> In the first argument of <E2><80><98>(.)<E2><80><99>, namely <E2><80><98>throwIO<E2><80><99>
      In the first argument of <E2><80><98>either<E2><80><99>, namely
        <E2><80><98>(throwIO . ExceptionNoSuchWallet)<E2><80><99>
      In the second argument of <E2><80><98>(>=>)<E2><80><99>, namely
        <E2><80><98>either (throwIO . ExceptionNoSuchWallet) pure<E2><80><99>
     |
2918 |     throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |                                        ^^^^^^^

src/Cardano/Wallet.hs:2918:73: error:
    <E2><80><A2> Could not deduce (Applicative m0) arising from a use of <E2><80><98>pure<E2><80><99>
      from the context: (AddressBookIso s, WalletFlavor s n k)
        bound by the type signature for:
                   delegationFee :: forall s (k :: Depth -> * -> *)
                                           (n :: Cardano.Wallet.Read.NetworkId.NetworkDiscriminant).
                                    (AddressBookIso s, WalletFlavor s n k) =>
                                    DBLayer IO s k
                                    -> NetworkLayer IO Read.Block
                                    -> TransactionLayer k 'CredFromKeyK SealedTx
                                    -> TimeInterpreter (ExceptT PastHorizonException IO)
                                    -> AnyRecentEra
                                    -> ChangeAddressGen s
                                    -> WalletId
                                    -> IO DelegationFee
        at src/Cardano/Wallet.hs:(2887,1)-(2899,23)
      or from: (MonadIO stm, MonadFail stm)
        bound by a pattern with constructor:
                   DBLayer :: forall (m :: * -> *) s (k :: Depth -> * -> *)
                                     (stm :: * -> *).
                              (MonadIO stm, MonadFail stm) =>
                              WalletId
                              -> (Wallet s
                                  -> WalletMetadata
                                  -> [(Tx, TxMeta)]
                                  -> GenesisParameters
                                  -> ExceptT
                                       Cardano.Wallet.DB.Errors.ErrWalletAlreadyInitialized stm ())
                              -> Data.DBVar.DBVar
                                   stm
                                   (WalletState.DeltaMap WalletId (WalletState.DeltaWalletState s))
                              -> (Wallet s -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (Wallet s))
                              -> stm [ChainPoint]
                              -> (WalletMetadata -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (WalletMetadata, WalletDelegation))
                              -> ExceptT ErrWalletNotInitialized stm Bool
                              -> (DelegationCertificate
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Coin -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm Coin
                              -> ([(Tx, TxMeta)] -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Maybe Coin
                                  -> SortOrder
                                  -> Range SlotNo
                                  -> Maybe TxStatus
                                  -> Maybe Natural
                                  -> stm [TransactionInfo])
                              -> (Hash "Tx"
                                  -> ExceptT ErrWalletNotInitialized stm (Maybe TransactionInfo))
                              -> (BuiltTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx"
                                  -> SealedTx -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm
                                   [Cardano.Wallet.DB.Store.Submissions.Operations.TxSubmissionsStatus]
                              -> (SlotNo
                                  -> [(SlotNo, Hash "Tx")]
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> (Hash "Tx" -> ExceptT ErrRemoveTx stm ())
                              -> ((k 'RootK XPrv, PassphraseHash)
                                  -> ExceptT ErrWalletNotInitialized stm ())
                              -> stm (Maybe (k 'RootK XPrv, PassphraseHash))
                              -> stm (Maybe GenesisParameters)
                              -> (Slot -> ExceptT ErrWalletNotInitialized stm ChainPoint)
                              -> (Quantity "block" GHC.Word.Word32
                                  -> SlotNo -> ExceptT ErrWalletNotInitialized stm ())
                              -> (forall a. stm a -> m a)
                              -> DBLayer m s k,
                 in an equation for <E2><80><98>delegationFee<E2><80><99>
        at src/Cardano/Wallet.hs:2900:18-28
      The type variable <E2><80><98>m0<E2><80><99> is ambiguous
      Relevant bindings include
        throwInIO :: ExceptT ErrNoSuchWallet m0 c -> m0 c
          (bound at src/Cardano/Wallet.hs:2918:5)
      These potential instances exist:
        instance Applicative (Either e) -- Defined in <E2><80><98>Data.Either<E2><80><99>
        instance Applicative NonEmpty -- Defined in <E2><80><98>GHC.Base<E2><80><99>
        instance Applicative IO -- Defined in <E2><80><98>GHC.Base<E2><80><99>
        ...plus 9 others
        ...plus 196 instances involving out-of-scope types
        (use -fprint-potential-instances to see them all)
    <E2><80><A2> In the second argument of <E2><80><98>either<E2><80><99>, namely <E2><80><98>pure<E2><80><99>
      In the second argument of <E2><80><98>(>=>)<E2><80><99>, namely
        <E2><80><98>either (throwIO . ExceptionNoSuchWallet) pure<E2><80><99>
      In the expression:
        runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |
2918 |     throwInIO = runExceptT >=> either (throwIO . ExceptionNoSuchWallet) pure
     |                                                                         ^^^^
[184 of 194] Compiling UnliftIO.Compat  ( src/UnliftIO/Compat.hs, dist/build/UnliftIO/Compat.o, dist/build/UnliftIO/Compat.dyn_o )
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[185 of 194] Compiling Cardano.Wallet.Shelley.Network.Node ( src/Cardano/Wallet/Shelley/Network/Node.hs, dist/build/Cardano/Wallet/Shelley/Network/Node.o, dist/build/Cardan>
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[186 of 194] Compiling Cardano.Wallet.Shelley.Network ( src/Cardano/Wallet/Shelley/Network.hs, dist/build/Cardano/Wallet/Shelley/Network.o, dist/build/Cardano/Wallet/Shelle>
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[187 of 194] Compiling Cardano.DB.Sqlite ( src/Cardano/DB/Sqlite.hs, dist/build/Cardano/DB/Sqlite.o, dist/build/Cardano/DB/Sqlite.dyn_o )
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[188 of 194] Compiling Cardano.Wallet.DB.Store.Checkpoints ( src/Cardano/Wallet/DB/Store/Checkpoints.hs, dist/build/Cardano/Wallet/DB/Store/Checkpoints.o, dist/build/Cardan>
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[189 of 194] Compiling Cardano.Wallet.DB.Sqlite.Stores ( src/Cardano/Wallet/DB/Sqlite/Stores.hs, dist/build/Cardano/Wallet/DB/Sqlite/Stores.o, dist/build/Cardano/Wallet/DB/>
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[190 of 194] Compiling Cardano.Wallet.DB.Sqlite.Migration ( src/Cardano/Wallet/DB/Sqlite/Migration.hs, dist/build/Cardano/Wallet/DB/Sqlite/Migration.o, dist/build/Cardano/W>
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[191 of 194] Compiling Cardano.Wallet.DB.Layer ( src/Cardano/Wallet/DB/Layer.hs, dist/build/Cardano/Wallet/DB/Layer.o, dist/build/Cardano/Wallet/DB/Layer.dyn_o )
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[192 of 194] Compiling Cardano.Pool.DB.Log ( src/Cardano/Pool/DB/Log.hs, dist/build/Cardano/Pool/DB/Log.o, dist/build/Cardano/Pool/DB/Log.dyn_o )
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
[193 of 194] Compiling Cardano.Pool.DB.Sqlite ( src/Cardano/Pool/DB/Sqlite.hs, dist/build/Cardano/Pool/DB/Sqlite.o, dist/build/Cardano/Pool/DB/Sqlite.dyn_o )
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
ld: warning: /nix/store/b3vd1djka4v8201x7j3590ldhssj7jxk-libiconv-50/lib/libiconv.dylib, ignoring unexpected dylib file
```

**Comment by Anviking** [comment]
GHC-8.10 wasn't happy about having the constraints as a superclass for some reason. I think dropping `ShelleyLedgerEra` should fix this. For now, I'm still going with one big `withConstraints` function, just without involving `IsRecentEra`.

**Pull request merged**